oeps/best-practices/oep-0018/decisions/0001-forbid-url-dependencies.rst

```diff
@@ -0,0 +1,287 @@
+  you do need to include a package at a URL, it should have both the package
+  name and version specified (end with "#egg=NAME==VERSION"). ...
+
+Issues with Former Guidance
+===========================
+
+Performance
+-----------
+
+Firstly, installing from VCS URL is significantly slower than installing from PyPI.
+We understand this in theory, based on the facts that:
+
+* ``pip`` must download and build a repository's entire contents when freshly installing
+  via VCS URL, whereas PyPI-hosted packages can be downloaded as pre-built wheels; and
+* when re-installing a dependencies, ``pip`` must fully re-install VCS-hosted
+  packages, whereas it will skip over any PyPI-hosted package whose version number has not
+  changed.
+
+We confirmed our theoretical understanding with a `cursory performance comparison`_.
+It demonstrated that installing a package into a fresh system from a VCS URL took
+nearly twice as a long as installing the same package from PyPI. It also
+demonstrated that installing an already-installed dependency from a VCS URL
+took four times as long as doing so for a package installed via PyPI.
+
+Upgrades
+--------
+
+OEP-18 describes a format for PyPI-hosted dependencies that allows for
+automated upgrades to be performed using the ``make uprade`` Makefile target.
+This target upgrades all version pins to the latest version, as allowed by the
+version restrictions specified in ``constraints.txt``.
+
+VCS URL dependencies work completely outside of this system. Each VCS URL
+is essentially a version pin, which cannot be upgraded by automated tooling.
+This runs counter to the "upgrade pins by default" philosophy of OEP-18.
+edx-platform's ``github.in`` file has VCS URL pins that have not been
+updated for over three years.
+
+Lack of reason to allow VCS URLs
+--------------------------------
+
+As noted above, OEP-18 carved out two exceptions in which VCS URLs were
+acceptable. Firstly:
+
+    You need to test a release candidate of the dependency to make sure it will
+    work with your code.
+
+This is a valid thing to test, but it should be done locally or in some sort of
+sandbox using a feature branch. A VCS URL should never be merged to main or master
+for testing purposes.
+
+Secondly:
+
+    You critically need a fix for a package which has not yet been included in
+    a release, and you cannot arrange for a release to be made in a timely
+    manner.
+
+Generally, fixes can be achieved by reverting to a previous, stable package version.
+In the rare event that reverting to a previous package version is not viable,
+Open edX repository maintainers can easily cut new releases of packages in order
+to accomodate a forward fix. If a repository maintainer is unresponsive, this
+issue can be escalated to tCRIL, which can cut an emergency release if necessary. If
+the repository is outside of the Open edX project, then the repository can be forked, and
+the package re-released to PyPI under a new name.
```

Comment:
One of the most common historical causes of VCS dependencies is needing support for a new Python or Django version for a named Open edX release before upstream is ready to merge and/or release it (because Open edX named releases come out before Python/Django EOL dates).  The absolute number is still small, but probably happens about 5 times across repos each Django upgrade cycle.

We can technically release forks of these to PyPI each time it happens, but just want to make clear that this will cause a fair bit of overhead each time it happens; we'd want a pretty solid workflow for this, and the "temporary" forks will pretty much sit in PyPI until we're willing to break all installations of the named release that needed it.